[WIP] Add experimental release metadata to module

tractorcow opened the pull request:
This will be used to allow better configurability of module releases in the future

In-progress readme at https://github.com/open-sausages/cow/blob/c951b7f000bbccf99ffb72c0d66609cccc8d2414/cow-schema.md

sminnee commented:
Can we merge this after the necessary cow changes are done?

sminnee commented:
Also, does it make sense to add type = recipe to composer.json as part of this?

sminnee commented:
Hrm, thinking about it more... Putting type = recipe on the composer.json will mean that projects created with it will also have that type, which isn't so good.

tractorcow commented:
type = recipe isn't needed anymore. A recipe is just a library that includes other libraries.

tractorcow commented:
Actually, type=recipe is needed for the plugin assets, sorry. The plugin will mis-install modules otherwise unless it's scoped (re the demo I gave that mistakenly installed the modules to vendor).

However, that's only necessary for the nested project files feature. :)

tractorcow commented:
I've added this line back in.

https://github.com/tractorcow/recipe-plugin/blob/master/src/RecipeInstaller.php#L22

Without this, the plugin scopes itself as "the" installer for all module types;

Possible alternative is to make this installer handle all silverstripe- prefixed types, and we actually deprecate the https://github.com/composer/installers/blob/master/src/Composer/Installers/SilverStripeInstaller.php installer.

tractorcow commented:
I've removed all self.version; This is how the composer should look for all dev branches of recipes. :)

Cow will now update the composer.json automatically on release (replace semver constraints with physical tags, and toggling prefer-stable to true).

tractorcow commented:
E.g. on tag of 4.0.0-alpha2, cow will update the composer.json to the below:

```
{
	"name": "silverstripe/installer",
	"description": "The SilverStripe Framework Installer",
	"require": {
		"php": ">=5.5.0",
		"silverstripe/cms": "4.0.0-alpha2",
		"silverstripe/framework": "4.0.0-alpha2",
		"silverstripe/reports": "4.0.0-alpha2",
		"silverstripe/siteconfig": "4.0.0-alpha2",
		"silverstripe-themes/simple": "~3.1.0",
		"silverstripe/asset-admin": "1.0.0-alpha2"
	},
	"require-dev": {
		"phpunit/PHPUnit": "~4.8"
	},
	"extra": {
		"branch-alias": {
			"dev-master": "4.0.x-dev"
		}
	},
	"config": {
		"process-timeout": 600
	},
	"prefer-stable": true,
	"minimum-stability": "dev"
}
```

sminnee commented on composer.json:
```diff
@@ -21,6 +21,6 @@
 		"process-timeout": 600
 	},
-	"prefer-stable": true,
+	"prefer-stable": false,
```
This will affect every package including all the symfony etc ones. Is there any other way of doing this?

tractorcow commented:
After some discussion we decided to leave prefer-stable to true, and I've added some more options to allow for customisation of tagged constraints.

By setting `"dependency-constraint": "allow-patch"` cow will tag the finished composer with `~4.0.0-alpha2`

tractorcow commented:
Support for these updates for cow are at https://github.com/open-sausages/cow/commit/d492091f9ba44a9e3daef5a8db5cd8b4015cd3e9

sminnee commented:
This seems fine to me, although I'm not sure how we'd test it. Have we documented how to use this to actually produce a new release yet?

dhensby commented:
Is the approach here that a recipe will lock people onto versions and they'll have to take manual intervention if they want semver constraints (eg `^4.0`)?

sminnee commented:
@dhensby right now, yes. It's the same as the current installer and without it, it wouldn't be possible to install a recipe that had anything other than the latest packages.

We may decide that it's better to install the latest patch releases of a minor release line and rewrite to less constrained dependencies. But I don't think that blocks merge.

The alternative approach would be to have some kind of "recipe upgrade" command the wouid rewrite composer dependencies if and when new recipe releases were out. This would be more controlled and would let us embargo recipe updates (e.g. if a recipe had an extra round of cross-module testing that may have not happened yet). However, it would be another tool to maintain and another composer command for users to remember. It's likely to have more use in special case recipes (such as CWP) than the default installer.

In any case, it's controlled by an option in the cow json.

tractorcow commented:
I've added a manual tests config, so that we can customise how recipes build tests.

tractorcow commented:
I've updated this in preparation for cow 2.0 release. Updated .cow.json schema at https://github.com/open-sausages/cow/blob/07d8efd4f4a1d5d7d440e27298fde17451291e50/cow-schema.md

tractorcow commented:
Cow 2.0 change is also open for review at https://github.com/silverstripe/cow/pull/21
